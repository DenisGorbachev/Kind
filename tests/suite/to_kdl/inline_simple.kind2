MyType (t: Type) (u: Type) : Type
MyType.new <t> <u> (a: U60) (b: U60) (c: t) (d: u) : MyType t u

MyType.swap <t> <u> (a: MyType t u) : MyType u t
MyType.swap (MyType.new a b c d) = MyType.new b a d c

MyType.mix <t> <u> (a: MyType t u) (b: MyType t u) : MyType t u
MyType.mix (MyType.new aa ab ac ad) (MyType.new ba bb bc bd) =
  MyType.new aa bb ac bd

MyType.get_c <t> <u> (a: MyType t u) : t
MyType.get_c (MyType.new a b c d) = c

#inline
SimpleFn <t> <u> (a: MyType t u) (b: MyType u t) : t {
  MyType.get_c (MyType.mix a (MyType.swap b))
}

Main {
  let a = (MyType.new 0 1 2 3)
  SimpleFn a ((x => MyType.new x 5 6 7) 4)
}