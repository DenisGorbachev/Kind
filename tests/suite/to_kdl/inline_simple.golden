// MyType.new -(t: Type) -(u: Type) (a: U60) (b: U60) (c: t) (d: u) : (MyType t u)
ctr {MyType_new a b c d}

// MyType.swap -(t: Type) -(u: Type) (a: (MyType t u)) : (MyType u t)
fun (MyType_swap a) {
  (MyType_swap {MyType_new a b c d}) =
    {MyType_new b a d c}
}

// MyType.mix -(t: Type) -(u: Type) (a: (MyType t u)) (b: (MyType t u)) : (MyType t u)
fun (MyType_mix a b) {
  (MyType_mix {MyType_new aa ~ ac ~} {MyType_new ~ bb ~ bd}) =
    {MyType_new aa bb ac bd}
}

// MyType.get_c -(t: Type) -(u: Type) (a: (MyType t u)) : t
fun (MyType_get_c a) {
  (MyType_get_c {MyType_new ~ ~ c ~}) =
    c
}

// Main : _
fun (Main) {
  (Main) =
    let a = {MyType_new #0 #1 #2 #3};
    (MyType_get_c (MyType_mix a (MyType_swap (!@x {MyType_new x #5 #6 #7} #4))))
}

